src/query/storage/converter.go

```diff
@@ -54,15 +55,30 @@ func PromWriteTSToM3(
 	}
 }
 
+// The default name for the name tag in Prometheus metrics.
+// This can be overwritten by setting tagOptions in the config
+var (
+	promDefaultName = []byte("__name__")
+)
+
 // PromLabelsToM3Tags converts Prometheus labels to M3 tags
 func PromLabelsToM3Tags(
 	labels []*prompb.Label,
 	tagOptions models.TagOptions,
 ) models.Tags {
 	tags := models.NewTags(len(labels), tagOptions)
-	tagList := make([]models.Tag, len(labels))
-	for i, label := range labels {
-		tagList[i] = models.Tag{Name: label.Name, Value: label.Value}
+	tagList := make([]models.Tag, 0, len(labels))
+	for _, label := range labels {
+		// If this label corresponds to the Prometheus name,
+		// instead set it as the given name tag from the config file.
+		if bytes.Equal(promDefaultName, label.Name) {
+			tags = tags.SetName(label.Value)
+		} else {
+			tagList = append(tagList, models.Tag{
```

Comment:
Since this slice is pre-allocated, anyway to index in to set the Tags?

Reply:
It gets a bit awkward since name will be added to Tags directly but not the slice, so you'd need to subtract from the length iff name tag is present otherwise you get nils